src/components/TodoContext/TodosContext.tsx

```diff
@@ -0,0 +1,381 @@
+      try {
+        const todosFromApi = await getTodos(user?.id);
+
+        setTodos(todosFromApi);
+      } catch {
+        setHidden(false);
+        setError('Unable to download todos');
+        throw new Error('Unable to download todos');
+      }
+    }
+  }
+
+  // adding todo when getting title from user
+  async function sendNewTodo(data: Todo) {
+    if (tempTodo) {
+      setTodos(current => (
+        [...current, { ...tempTodo, title: tempTodo.title.trim() }]
+      ));
+    }
+
+    if (user && data) {
+      try {
+        const todoFromServer = await sendTodo(user?.id, data);
+
+        setTodos(current => current.map(todo => {
+          if (todo.id === 0) {
+            return { ...todo, id: todoFromServer.id };
+          }
+
+          return todo;
+        }));
+      } catch {
+        setHidden(false);
+        setTodos(current => current.slice(0, current.length - 1));
+        setError('Unable to add a todo');
+        throw new Error('Unable to add a todo');
+      }
+    }
+
+    setNewTodo(null);
+  }
+
+  // delete todos or todo
+  async function deleteTodos(todosId: number[]) {
+    if (user && todosId) {
+      try {
+        await Promise.all(todosId.map(todoId => deleteTodo(todoId)));
+
+        setTodos(currentTodos => (
+          currentTodos.filter(({ id }) => !todosId.includes(id))
+        ));
+        setModifiedTodosId([0]);
+      } catch {
+        setModifiedTodosId([0]);
+        setHidden(false);
+        setError('Unable to delete a todo');
+        throw new Error('Unable to delete a todo');
+      }
+    }
+  }
+
+  // modify todos through modifying complete status or title
+  async function modifyTodos(changedTodos: Todo[]) {
+    if (user) {
+      try {
+        const todosFromServer = await Promise
+          .all(changedTodos.map(todo => patchTodo(todo)));
+
+        setTodos(currentTodos => currentTodos.map(todo => {
+          const modifiedTodo: Todo | null = todosFromServer
+            .find(todoFromServer => todoFromServer.id === todo.id) || null;
+
+          if (modifiedTodo) {
+            return {
+              ...todo,
+              title: modifiedTodo.title || '',
+              completed: modifiedTodo.completed || false,
+            };
+          }
+
+          return todo;
+        }));
+        setModifiedTodosId([0]);
+      } catch {
+        setModifiedTodosId([0]);
+        setHidden(false);
+        setError('Unable to update a todo');
+        throw new Error('Unable to update a todo');
+      }
+    }
+  }
+
+  // #endregion
+
+  // #region ---- HEADER ----
+  // handle submit action of main input (from where title comes)
+  function handleNewSubmit(event: React.FormEvent<HTMLFormElement>) {
+    event.preventDefault();
+    if (newTodo?.title.trim()) {
+      sendNewTodo(newTodo);
+    } else {
+      setHidden(false);
+      setNewTodo(null);
+      setError('Title can\'t be empty or consist only spaces');
+    }
+
+    newTodoField.current?.blur();
+  }
```

Comment:
I don't think this function should be in TodosContext, all methods that belong to the form should be in the component with the form